File: src/19-challenge/solution.js
/**
 * @param {number} weight - Total weight of the gifts
 * @returns {string} - Stacked boxes represented as ASCII art
 */
export default function distributeWeight(weight) {
	/** @type {Object.<number, string[]>} */
	const boxRepresentations = {
		1: [' _ ', '|_|'],
		2: [' ___ ', '|___|'],
		5: [' _____ ', '|     |', '|_____|'],
		10: [' _________ ', '|         |', '|_________|'],
	}

	/** @type {string[][]} */
	const boxes = []

	/** @type {string[]} */
	const weights = Object.keys(boxRepresentations)

	for (let i = weights.length - 1; i > -1; i--) {
		/** @type {number} */
		const boxWeight = parseInt(weights[i])
		if (weight - boxWeight < 0) continue

		weight -= boxWeight
		boxes.unshift(boxRepresentations[boxWeight])

		while (weight - boxWeight > -1) {
			weight -= boxWeight
			boxes.unshift(boxRepresentations[boxWeight])
		}

		if (!weight) break
	}

	/** @type {string[]} */
	const boxesToDraw = []

	for (let i = 0; i < boxes.length; i++) {
		/** @type {string[]} */
		const box = boxes[i]

		if (i !== boxes.length - 1) {
			/** @type {number} */
			const offset = ~~`${boxes[i + 1][0].trim().match(/_/g)?.length}` + 1
			box[box.length - 1] = box[box.length - 1].padEnd(offset, '_')
		}

		boxesToDraw.push(box.slice(~~!!~~`${i}`).join('\n'))
	}

	return boxesToDraw.join('\n')
}

File: src/19-challenge/solution.py
def distribute_weight(weight: int) -> str:
    box_representations: dict[int, list[str]] = {
        1: [" _ ", "|_|"],
        2: [" ___ ", "|___|"],
        5: [" _____ ", "|     |", "|_____|"],
        10: [" _________ ", "|         |", "|_________|"],
    }

    boxes: list[list[str]] = []
    weights: list[int] = list(box_representations.keys())

    for i in range(len(weights) - 1, -1, -1):
        box_weight: int = weights[i]
        if weight - box_weight < 0:
            continue

        weight -= box_weight
        boxes.insert(0, box_representations[box_weight])

        while weight - box_weight > -1:
            weight -= box_weight
            boxes.insert(0, box_representations[box_weight])

        if weight == 0:
            break

    boxes_to_draw: list[str] = []

    for i, box in enumerate(iterable=boxes):
        if i != len(boxes) - 1:
            offset: int = boxes[i + 1][0].strip().count("_") + 1
            box[-1] = box[-1].ljust(offset, "_")

        if i != 0:
            boxes_to_draw.append("\n".join(box[1:]))
            continue

        boxes_to_draw.append("\n".join(box))

    return "\n".join(boxes_to_draw)
